solc/CommandLineParser.cpp

```diff
@@ -967,7 +967,8 @@ General Information)").c_str(),
 		m_options.optimizer.enabled = (m_args.count(g_strOptimize) > 0);
 		m_options.optimizer.noOptimizeYul = (m_args.count(g_strNoOptimizeYul) > 0);
 
-		m_options.optimizer.expectedExecutionsPerDeployment = m_args.at(g_strOptimizeRuns).as<unsigned>();
+		if (m_args.count(g_strOptimizeRuns))
```

Comment:
Isn't this always true? It has a default value.
https://github.com/ethereum/solidity/blob/develop/solc/CommandLineParser.cpp#L693

Reply:
I think it would be better to let the default value be defined through the consturctors of `OptimizerSettings`

Reply:
I meant, the condition `m_args.count(g_strOptimizeRuns)` is always true, because we set a default value in line https://github.com/ethereum/solidity/blob/develop/solc/CommandLineParser.cpp#L693

Reply:
Yes, I know, so the PR was a correct, but since we now the commandline parser result is an `optional`, we should also make use of that. Essentially: If the option was not given, then use whatever the optiimzer setting constructor does, and if it is set, modify it in the optimizer settings afterwards.

Reply:
I agree with @hrkrshnn here - this change is basically a no-op. The condition is always true because of the default so the `if` could just as well be an assert.

If would make sense if we removed `->default_value()` but the nice thing about having it is that the value is visible in the command-line help. Removing it would make sense if the default was dynamic or different for each optimizer level but that's not the case here so I'd be for keeping it. At least after #11550 the default is not hard-coded and comes from `OptimiserSettings`.

Reply:
My main worry is that if we at some point introduce e.g. `OptimiserSettings::size()` and `OptimiserSettings::speed()`, it will not work because the commandline interface always sets "runs" to 200.

Reply:
By the way, a lot more settings in `CommandLineOptions` should really be `optional` and aren't only because I did not want to introduce that change in the big refactor PR. In some cases it was hard to know if the default is actually used or if it just represents an "uninitialized" state is never used so I left it for a smaller PR that can be properly reviewed.

Reply:
I created a task for converting settings to optionals: #11727. I thought there would be more of them to convert but looks like it's just a few. It's an easy refactor so I marked as `good first issue` and added a detailed description - hopefully some contributor picks that up.

> My main worry is that if we at some point introduce e.g. `OptimiserSettings::size()` and `OptimiserSettings::speed()`, it will not work because the commandline interface always sets "runs" to 200.

The question is if we want to make the command-line help a tiny bit less useful now because of that. Personally, I'd rather keep the default in command-line help because someone introducing new presets specifically with a different runs value is expected to test them and should easily notice that the parameter gets the wrong value.

To make it even less likely to be overlooked we could even add an assert in `CommandLineInterface` after optimizer settings struct is initialized but before `expectedExecutionsPerDeployment` field is assigned - it would fail if the value of the field is different than the default and have a message to update the command-line default.

But yeah, that's not 100% foolproof so removing `>default_value()` and just having it at `nullopt` is also fine with me. Both ways have some downsides.

Reply:
We should certainly display the default value, but can't we distinguish between "not given" and "--runs 200"? I would like to to modify the value in the first case - please also see the command line options: the field was changed to an `optional`, and we should either make use of the optional or revert this change.

Reply:
ok, that would be the third way to do it. In that case you just need to adjust the condition so that it actually distinguishes between default being used or not:
```suggestion
		if (!m_args[g_strOptimizeRuns].defaulted())
```

Reply:
I'll add this to the PR.